low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -22,28 +22,227 @@
+                    break
+        if not fru_found:
+            raise ResourceMapError(
+                errno.EINVAL,
+                "Health provider doesn't have support for'{rpath}'.")
+        return info
+
+    @staticmethod
+    def get_node_details(node):
+        """
+        Parse node information and returns left string and instance.
+
+        Example
+            "storage"    -> ("storage", "*")
+            "storage[0]" -> ("storage", "0")
+        """
+        res = re.search(r"(\w+)\[([\d]+)\]|(\w+)", node)
+        inst = res.groups()[1] if res.groups()[1] else "*"
+        node = res.groups()[0] if res.groups()[1] else res.groups()[2]
+        return node, inst
+
+    def get_data_template(self, uid, is_fru: bool):
+        return {
+            "uid": uid,
+            "fru": str(is_fru).lower(),
+            "last_updated": int(time.time()),
+            "health": {
+                "status": "",
+                "description": "",
+                "recommendation": "",
+                "specifics": []
+            }
+        }
+
+    def set_health_data(self, obj: dict, status, description=None,
+                        recommendation=None):
+        good_state = (status == "OK")
+        if not description:
+            description = "%s is %s in good state" % (
+                            obj.get("uid"),
+                            '' if good_state else 'not')
+        if not recommendation:
+            recommendation = 'None' if good_state\
+                             else "Contact Seagate Support."
+
+        obj["health"].update({
+            "status": status,
+            "description": description,
+            "recommendation": recommendation
+        })
+
+    def get_sas_hba_info(self):
+        return self.get_sas_ports_info("sas_hba")
+
+    def get_sas_ports_info(self, leaf_node="sas_ports"):
+        sas_hba_data = []
+        health_data = []
+        for sas_rid in self.SAS_RESOURCE_ID:
+            r_data = self.get_data_template(sas_rid, False)
+            sysfs_instance = ToolFactory().get_instance('sysfs')
+            sysfs_instance.initialize()
+            sas_ports_dict = sysfs_instance.get_phy_negotiated_link_rate()
+            sas_ports_dict = sorted(sas_ports_dict.items(),
+                                    key=lambda kv: int(kv[0].split(':')[1]))
+            health = []
+            if sas_ports_dict:
+                for rid, rate in sas_ports_dict:
+                    r_data["health"]["specifics"].append({
+                        "resource_id": f"{sas_rid}:{rid.strip()}",
+                        "negotiated_link_rate": rate.strip()
+                    })
+                    if "Gbit".lower() in rate.strip().lower():
+                        health.append("OK")
+
+            # By some logic set health status, description, recommendation
+            health_data.append(health)
+            health_status = "OK" if "OK" in health else "Fault"
```

Comment:
Hi @satish-seagate, currently we have this logic where, we declare "OK" status for sas_hba when any one of the phy link has negotiated link rate in Gbit. Is this logic okay? According to your sample health data, since it has one link down the status for sas_hba is shown as "Fault". So what should be the logic for health status of sas_hba??

Reply:
@sumedhak27 both conditions derive the SAS HBA status. First all links should be up & their link rate matters. So if link is Up, and then look for Gbit logic ..